### PR TITLE
Cogvideox_fun_inp pipeline

### PR DESCRIPTION
**Pull request by satani99**
Added CogvideoxFunInpaintPipeline 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sayakpaul

### CODE REVIEW

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 1]
```diff
@@ -0,0 +1,1057 @@
...
+else:
+    XLA_AVAILABLE = False
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```python
+        >>> import torch
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:52]:
```suggestion
        >>> import PIL.Image
        >>> import torch
```
As we use `PIL.Image` below when creating `mask_video`.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+        >>> pipe = CogVideoXFunInpaintPipeline.from_pretrained(
+        ...     "alibaba-pai/CogVideoX-Fun-V1.1-5b-InP", torch_dtype=torch.bfloat16
+        ... )
+        >>> pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)
+        >>> pipe.to("cuda")
+
+        >>> video = load_video(
+        ...     "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/hiker.mp4"
+        ... )
+        >>> mask_video = [Image.new("L", frame.size, 255) for frame in video]
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:65]:
```suggestion
        >>> mask_video = [PIL.Image.new("L", frame.size, 255) for frame in video]
```
Follow up change to https://github.com/huggingface/diffusers/pull/13331#discussion_r2991990954.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+        >>> prompt = "A cinematic mountain hike with dramatic lighting."
+
+        >>> output = pipe(prompt=prompt, video=video, mask_video=mask_video, output_type="pt").frames[0]
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:68]:
```suggestion
        >>> output = pipe(prompt=prompt, video=video, mask_video=mask_video, output_type="np").frames[0]
```
`export_to_video` only accepts `list[np.ndarray]` or `list[PIL.Image.Image]` video arguments.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 4]
```diff
@@ -0,0 +1,1057 @@
...
+                prompt=negative_prompt,
+                num_videos_per_prompt=num_videos_per_prompt,
+                max_sequence_length=max_sequence_length,
+                device=device,
+                dtype=dtype,
+            )
+
+        return prompt_embeds, negative_prompt_embeds
+
+    def _preprocess_mask_video(self, mask_video, height: int, width: int) -> torch.Tensor:
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:370]:
```suggestion
    @staticmethod
    def _preprocess_mask_video(mask_video, height: int, width: int) -> torch.Tensor:
```
Since `_preprocess_mask_video` does not use the `self` argument

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 5]
```diff
@@ -0,0 +1,1057 @@
...
+        if return_video_latents or (latents is None and not is_strength_max):
+            video = video.to(device=device, dtype=self.vae.dtype)
+
+            videos = []
+            for i in range(video.size(0)):
+                video_bs = video[i : i + 1]
+                video_bs = self.vae.encode(video_bs)[0]
+                video_bs = video_bs.sample()
+                videos.append(video_bs)
+            video = torch.cat(videos, dim=0)
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:429]: I think we should encode the `video` once instead of batch element by batch element, as this would better respect sliced/tiled encoding settings on the VAE. Users can always specify tiled encoding via `pipe.vae.enable_tiling`.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 6]
```diff
@@ -0,0 +1,1057 @@
...
+            noise = latents.to(device)
+            latents = noise * self.scheduler.init_noise_sigma
+
+        outputs = (latents,)
+
+        if return_noise:
+            outputs += (noise,)
+
+        if return_video_latents:
+            outputs += (video_latents,)
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:450]: I think we should remove the `return_noise` and `return_video_latents` arguments because these extra outputs are not used by `CogVideoXFunInpaintPipeline`.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 7]
```diff
@@ -0,0 +1,1057 @@
...
+                masked_image = add_noise_to_reference_video(masked_image, ratio=noise_aug_strength)
+
+            masked_image = masked_image.to(device=device, dtype=self.vae.dtype)
+            mask_pixel_values = []
+            for i in range(masked_image.size(0)):
+                mask_pixel_value = masked_image[i].unsqueeze(0)
+                mask_pixel_value = self.vae.encode(mask_pixel_value)[0]
+                mask_pixel_value = mask_pixel_value.mode()
+                mask_pixel_values.append(mask_pixel_value)
+            masked_image_latents = torch.cat(mask_pixel_values, dim=0)
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:483]: Similar to https://github.com/huggingface/diffusers/pull/13331#discussion_r2992010172, I think we should encode `masked_image` once here to respect any slicing/tiling enabled on the VAE.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 8]
```diff
@@ -0,0 +1,1057 @@
...
+            masked_image_latents = masked_image_latents * self.vae_scaling_factor_image
+        else:
+            masked_image_latents = None
+
+        return mask, masked_image_latents
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:488]:
```suggestion
        return masked_image_latents
```
I think we should remove the `mask` return value here as it is not used in `__call__`.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 9]
```diff
@@ -0,0 +1,1057 @@
...
+                raise ValueError(
+                    "`prompt_embeds` and `negative_prompt_embeds` must have the same shape when passed directly, but"
+                    f" got: `prompt_embeds` {prompt_embeds.shape} != `negative_prompt_embeds`"
+                    f" {negative_prompt_embeds.shape}."
+                )
+
+        if video is not None and latents is not None:
+            raise ValueError("Only one of `video` or `latents` should be provided.")
+
+    def fuse_qkv_projections(self) -> None:
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:575]: I think we should remove the `fuse_qkv_projections`/`unfuse_qkv_projections` pipeline methods in favor of calling the corresponding methods directly on the transformer (e.g. `pipe.transformer.fuse_qkv_projections()`), similar to what we do for VAE slicing/tiling methods. CC @yiyixuxu

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 10]
```diff
@@ -0,0 +1,1057 @@
...
+
+    @property
+    def current_timestep(self):
+        return self._current_timestep
+
+    @property
+    def interrupt(self):
+        return self._interrupt
+
+    def get_timesteps(self, num_inference_steps, timesteps, strength, device):
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:651]: I think the code would be more clear if we inlined the logic in the `get_timesteps` method in `__call__`. We generally prefer inlining smaller methods into the main pipeline code.

File: src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py [hunk 11]
```diff
@@ -0,0 +1,1057 @@
...
+            `tuple`. When returning a tuple, the first element is a list with the generated images.
+        """
+        if isinstance(callback_on_step_end, (PipelineCallback, MultiPipelineCallbacks)):
+            callback_on_step_end_tensor_inputs = callback_on_step_end.tensor_inputs
+
+        height = height or self.transformer.config.sample_height * self.vae_scale_factor_spatial
+        width = width or self.transformer.config.sample_width * self.vae_scale_factor_spatial
+        num_frames = num_frames or self.transformer.config.sample_frames
+        
+        num_videos_per_prompt = 1
```
Comment (dg845) [inline src/diffusers/pipelines/cogvideo/pipeline_cogvideox_fun_inpaint.py:791]: I think we should support `num_videos_per_prompt > 1` unless there is a strong reason not to.
